Wrappers/Python/cil/optimisation/utilities/StepSizeMethods.py

```diff
@@ -0,0 +1,136 @@
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+# Authors:
+# - CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
+
+from abc import ABC, abstractmethod
+import numpy
+
+
+class StepSizeRule(ABC):
+    """
+    Abstract base class for a step size rule. The abstract method, `get_step_size` takes in an algorithm and thus can access all parts of the algorithm (e.g. current iterate, current gradient, objective functions etc) and from this  should return a float as a step size. 
+    """
+
+    def __init__(self):
+        '''Initialises the step size rule 
+        '''
+        pass
+
+    @abstractmethod
+    def get_step_size(self, algorithm):
+        """
+        Returns
+        --------
+        float: the calculated step size 
+        """
+        pass
+
+
+class ConstantStepSize(StepSizeRule):
+    """
+    Step-size rule that always returns a constant step-size. 
+
+    Parameters
+    ----------
+    step_size: float
+        The step-size to be returned with each call. 
+    """
+
+    def __init__(self, step_size):
+        '''Initialises the step size rule 
+        '''
+        self.step_size = step_size
+
+    def get_step_size(self, algorithm):
+        """
+        Returns
+        --------
+        float: the calculated step size 
+        """
+        return self.step_size
+
+
+class ArmijoStepSizeRule(StepSizeRule):
+
+    """ 
+    Applies the Armijo rule to calculate the step size (step_size).
+
+    The Armijo rule runs a while loop to find the appropriate step_size by starting from a very large number (`alpha`). The step_size is found by reducing the step size (by a factor `beta`) in an iterative way until a certain criterion is met. To avoid infinite loops, we add a maximum number of times (`kmax`) the while loop is run.
+
+    Parameters
+    ----------
+    alpha: float, optional, default=1e6
+        The starting point for the step size iterations 
+    beta: float between 0 and 1, optional, default=0.5
+        The amount the step_size is reduced if the criterion is not met
+    kmax: integer, optional, default is numpy.ceil (2 * numpy.log10(alpha) / numpy.log10(2))
+        The maximum number of iterations to find a suitable step size 
+
+    Reference
+    ---------
+    Algorithm 3.1 (Numerical Optimization, Nocedal, Wright) (https://www.math.uci.edu/~qnie/Publications/NumericalOptimization.pdf)
+     https://projecteuclid.org/download/pdf_1/euclid.pjm/1102995080
+
+    """
+
+    def __init__(self, alpha=None, beta=None, kmax=None):
```

Comment:
```suggestion
    def __init__(self, alpha=1e6, beta=0.5, kmax=None):
```

This should also simplify the code below